phoenicis-engines/src/main/java/org/phoenicis/engines/EngineSettingsManager.java

```diff
@@ -19,98 +19,128 @@
 
-import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.ExecutorService;
 import java.util.function.Consumer;
+import java.util.stream.Collectors;
 
 /**
- * manages the engine settings
+ * Manages the engine settings
  */
 public class EngineSettingsManager {
-    private final ScriptInterpreter scriptInterpreter;
+    private final PhoenicisScriptEngineFactory phoenicisScriptEngineFactory;
+    private final ExecutorService executorService;
 
     /**
-     * constructor
-     * @param scriptInterpreter
+     * Constructor
+     *
+     * @param phoenicisScriptEngineFactory The used script engine factory
+     * @param executorService The executor service to allow for parallelization
      */
-    public EngineSettingsManager(ScriptInterpreter scriptInterpreter) {
-        this.scriptInterpreter = scriptInterpreter;
-    }
+    public EngineSettingsManager(PhoenicisScriptEngineFactory phoenicisScriptEngineFactory,
+            ExecutorService executorService) {
+        super();
 
-    /**
-     * fetches the required setting
-     * @param engineId setting ID (e.g. "glsl")
-     * @param settingId setting ID (e.g. "glsl")
-     * @param doneCallback callback which will be executed with the fetched setting
-     * @param errorCallback callback which will be executed if an error occurs
-     */
-    public void getSetting(String engineId, String settingId, Consumer<EngineSetting> doneCallback,
-            Consumer<Exception> errorCallback) {
-        final InteractiveScriptSession interactiveScriptSession = scriptInterpreter.createInteractiveSession();
-
-        interactiveScriptSession.eval(
-                "include(\"engines." + engineId + ".settings." + settingId + "\"); new Setting();",
-                output -> {
-                    final EngineSetting setting = ((Value) output).as(EngineSetting.class);
-                    doneCallback.accept(setting);
-                }, errorCallback);
+        this.phoenicisScriptEngineFactory = phoenicisScriptEngineFactory;
+        this.executorService = executorService;
     }
 
     /**
-     * fetches the available engine settings
-     * @param repositoryDTO
-     * @param callback
-     * @param errorCallback callback which will be executed if an error occurs
+     * Fetches the available engine settings
+     *
+     * @param repositoryDTO The repository containing the engine settings
+     * @param callback The callback which recieves the found engine settings
+     * @param errorCallback The callback which will be executed if an error occurs
      */
     public void fetchAvailableEngineSettings(RepositoryDTO repositoryDTO,
             Consumer<Map<String, List<EngineSetting>>> callback, Consumer<Exception> errorCallback) {
-        final InteractiveScriptSession interactiveScriptSession = scriptInterpreter.createInteractiveSession();
+        executorService.execute(() -> {
```

Comment:
Just as a note:
I plan to make the usage of an async environment when evaluating scripts explicit (like I did here).
I would like to hear your thoughts on this @qparis and @plata

Reply:
I'm not quite sure I understand you correctly. Could you explain in more detail?